Add cleartext traffic check

akindofyoga opened the pull request:
When someone tries to open a Websocket connection without using TLS (ie "ws:" instead of "wss:") on Android 9 and newer, the connection will fail and Scarlet will not provide any information about this to the developer. This confused me as well as the author of [this](https://stackoverflow.com/q/57675908/859277) post. 

Okhttp has a similar check [here](https://github.com/square/okhttp/blob/7a641531c8e03b10dc4fbf9f2465775d7ecc7649/okhttp/src/main/java/okhttp3/internal/connection/RealConnection.kt#L160). 

I check for a ClassNotFoundException, so it should not cause any issues for non-Android apps that use Scarlet.

akindofyoga committed:
Add cleartext traffic check

akindofyoga committed:
Update OkHttpClientUtils.kt

akindofyoga committed:
Update build.gradle

akindofyoga committed:
Update OkHttpClientUtils.kt

akindofyoga committed:
Fixed dependency

aaronhe42 commented on scarlet-websocket-okhttp/src/main/java/com/tinder/scarlet/websocket/okhttp/OkHttpClientUtils.kt:
```diff
@@ -16,5 +18,16 @@ fun OkHttpClient.newWebSocketFactory(requestFactory: RequestFactory): WebSocket.
+        try {
+            if (Build.VERSION.SDK_INT >= 23 &&
+                    !NetworkSecurityPolicy.getInstance().isCleartextTrafficPermitted()) {
+                throw RuntimeException("Android configuration does not permit cleartext traffic.")
```
the error type and message in okhttp sound better `UnknownServiceException` with `CLEARTEXT communication to $host not permitted by network security policy`, and

rogerthat94 replied:
Did part of this comment get cut off?

aaronhe42 commented on scarlet-websocket-okhttp/src/main/java/com/tinder/scarlet/websocket/okhttp/OkHttpClientUtils.kt:
```diff
@@ -16,5 +18,16 @@ fun OkHttpClient.newWebSocketFactory(requestFactory: RequestFactory): WebSocket.
+    if (url.startsWith("ws://")) {
+        try {
+            if (Build.VERSION.SDK_INT >= 23 &&
+                    !NetworkSecurityPolicy.getInstance().isCleartextTrafficPermitted()) {
```
we'll need to support the overload method added in 24 as well. see: https://developer.android.com/reference/android/security/NetworkSecurityPolicy.html#isCleartextTrafficPermitted(java.lang.String). but this would need a little refactor to pass in host name.

rogerthat94 replied:
OkHttp has a fairly complex process to get the [host](https://github.com/square/okhttp/blob/4b5615c3323d8abdb875f34a6cd9292825308faf/okhttp/src/main/java/okhttp3/HttpUrl.kt#L1324). But it seems like [this](https://developer.android.com/reference/java/net/URL.html#getHost()) method should suit our purposes.

aaronhe42 commented on scarlet-websocket-okhttp/src/main/java/com/tinder/scarlet/websocket/okhttp/OkHttpClientUtils.kt:
```diff
@@ -16,5 +18,16 @@ fun OkHttpClient.newWebSocketFactory(requestFactory: RequestFactory): WebSocket.
 }
 
 fun OkHttpClient.newWebSocketFactory(url: String): WebSocket.Factory {
+    if (url.startsWith("ws://")) {
```
dont think this check will be necessary

rogerthat94 replied:
We need this check because people using SSL don't need to have cleartext traffic permitted. OkHttp does this check with the combination of code [here](https://github.com/square/okhttp/blob/20cd3a08935c78c75dae015964608b2c233d629d/okhttp/src/main/java/okhttp3/Request.kt#L175), [here](https://github.com/square/okhttp/blob/7a641531c8e03b10dc4fbf9f2465775d7ecc7649/okhttp/src/main/java/okhttp3/Address.kt#L75) and [here](https://github.com/square/okhttp/blob/7a641531c8e03b10dc4fbf9f2465775d7ecc7649/okhttp/src/main/java/okhttp3/internal/connection/RealConnection.kt#L154). I have added the "ignoreCase = true" parameter to String.startsWith that OkHttp uses.

aaronhe42 commented:
thanks for the contribution @rogerthat94! Just a few nits

akindofyoga committed:
Requested changes

rogerthat94 commented:
Thank you for making this helpful library! I'm happy to give back with a small contribution. 

I hope this addresses the nits.

akindofyoga committed:
Change substring check

akindofyoga committed:
Fix hostname

akindofyoga committed:
Change url passed to java.net.URL to start with `http:`.

It throws an exception when the URL starts with `ws:.

akindofyoga committed:
Use okhttp3.HttpUrl instead of java.net.URL

aaronweihe commented:
@rogerthat94 sorry it's been a while. if you can resolve conflicts, i'd be more than happy to accept this change. Thanks!

akindofyoga committed:
Merge branch 'master' into Cleartext-Traffic

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/Tinder/Scarlet?pullRequest=122) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/Tinder/Scarlet?pullRequest=122) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/Tinder/Scarlet?pullRequest=122) it.</sub>

rogerthat94 commented:
@aaronweihe No problem! I have just resolved conflicts.

akindofyoga committed:
Update dependencies.gradle

akindofyoga committed:
Update dependencies.gradle

akindofyoga committed:
Update dependencies.gradle

akindofyoga committed:
Update build.gradle

akindofyoga committed:
Fix Android dependency.

It looks like the version of this cannot be specified separately

OkHttp specifies this in a single string. See here:
https://github.com/square/okhttp/blob/68241851c197a67cdcdd0e150e239f55d1e556f4/build.gradle#L28

rogerthat94 commented:
ping @aaronweihe

akindofyoga commented:
@aaronweihe Are you still interested in this change? I can rebase it on top of the latest version of the code if so.

aaronweihe reviewed (approved):
looks good to me. Can you sign the CLA?

aaronweihe commented on scarlet-websocket-okhttp/src/main/java/com/tinder/scarlet/websocket/okhttp/OkHttpClientUtils.kt:
```diff
@@ -6,15 +6,34 @@
 fun OkHttpClient.newWebSocketFactory(url: String): WebSocket.Factory {
+    if (url.startsWith("ws:", ignoreCase = true)) {
+        try {
+            if ((Build.VERSION.SDK_INT > 23 &&
```
I like the [`Platform#isCleartextTrafficPermitted`](https://github.com/square/okhttp/blob/master/okhttp/src/jvmMain/kotlin/okhttp3/internal/platform/Platform.kt#L136) abstraction. we could tackle it in the next PR

akindofyoga commented:
Thanks! I have signed the CLA. Is there anything else that you would like me to do?

akindofyoga commented:
@aaronweihe Are you able to merge this?
